### PR TITLE
Refactor: use StepRws in evm circuits to replace rw indices 

### PR DESCRIPTION
**Pull request by maschad**
### Description

Continuing the work originally started in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1616 this PR seeks to introduce `StepRws` across the evm circuits to reduce the need for tracking rw indices through code.

### Issue Link
Closes https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1586

### Type of change
Refactor (no updates to logic)

### Contents

The change has been made to the following files

 - error_oog_call.rs
 - error_oog_memory_copy.rs
 - error_oog_sload_sstore.rs
 - error_return_data_oo_bound.rs
 - error_write_protection.rs
 - extcodecopy.rs
 - extcodehash.rs
 - extcodesize.rs
 - logs.rs
 - return_revert.rs
 - returndatacopy.rs
 - sha3.rs
 - sload.rs
 - sstore.rs

### CODE REVIEW

**Comment by maschad** [comment]
@ChihChengLiang this should be ready for review now

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs [hunk 1]
```diff
@@ -127,26 +127,39 @@ impl<F: Field> ExecutionGadget<F> for ErrorOOGCallGadget<F> {
...
 
-        let callee_code_hash = block
-            .get_rws(step, 9 + is_call_or_callcode)
-            .account_codehash_pair()
-            .0;
+        if is_call_or_callcode == 1 {
+            rws.offset_sub(1);
+        }
+
+        rws.offset_add(is_call_or_callcode);
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:149]: Isn't the following block unnecessary?
```
if is_call_or_callcode == 1 {
    rws.offset_sub(1);
}

rws.offset_add(is_call_or_callcode);
```

If `is_call_or_callcode == 1` you first decrease by 1 and then increase by 1.
If `is_call_or_callcode == 0` you don't decrease by 1 (stays the same) and then increase by 0 (stays the same).

File: zkevm-circuits/src/evm_circuit/execution/sload.rs [hunk 2]
```diff
@@ -121,16 +121,25 @@ impl<F: Field> ExecutionGadget<F> for SloadGadget<F> {
...
-        let value = block.get_rws(step, 6).stack_value();
+        let mut rws = StepRws::new(block, step);
+
+        rws.offset_add(4);
+
+        let key = rws.next().stack_value();
+
+        let (_, committed_value) = rws.next().aux_pair();
+
+        let value = rws.next().stack_value();
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/execution/sload.rs:132]: Nitpick: perhaps remove the empty lines in between `rws.next` calls (also in other places)?

**Review by miha-stopar — APPROVED** [review]
LGTM! Just two nitpicks.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs (hunk 1 above)
Comment (maschad) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs:149]: My bad 😅  looks like some debugging residue, Good catch!

**Comment by maschad** [comment]
Thanks for the review @miha-stopar !  I've made the changes.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs [hunk 3]
```diff
@@ -165,17 +167,22 @@ impl<F: Field> ExecutionGadget<F> for ErrorOOGMemoryCopyGadget<F> {
...
         } else {
             (false, U256::zero())
         };
 
-        let rw_offset = if is_extcodecopy { 3 } else { 0 };
-        let [dst_offset, src_offset, copy_size] = [rw_offset, rw_offset + 1, rw_offset + 2]
-            .map(|index| block.get_rws(step, index).stack_value());
+        if !is_extcodecopy {
+            rws.offset_reset();
+        };
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs:181]: This line is not required.
If `is_extcodecopy` is false, then the three `rws.next()` in line 170~173 wouldn't have happened. rw_offset here would be still zero.

From the API perspective, rw_index represents the order or reads and writes. It should be monotonic increasing and we shouldn't design an API to make it decrease or reset to zero.
```suggestion
```

File: zkevm-circuits/src/evm_circuit/execution/returndatacopy.rs [hunk 4]
```diff
@@ -187,8 +188,8 @@ impl<F: Field> ExecutionGadget<F> for ReturnDataCopyGadget<F> {
             (4, CallContextFieldTag::LastCalleeReturnDataOffset),
             (5, CallContextFieldTag::LastCalleeReturnDataLength),
         ]
-        .map(|(i, tag)| {
-            let rw = block.get_rws(step, i);
+        .map(|(_, tag)| {
+            let rw = rws.next();
             assert_eq!(rw.field_tag(), Some(tag as u64));
             rw.call_context_value()
         });
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/execution/returndatacopy.rs:195]: This whole block from line 185 can be replaced with:
```
let last_callee_id = rws.next().call_context_value();
let return_data_offset = rws.next().call_context_value();
let return_data_size = rws.next().call_context_value();
```

File: zkevm-circuits/src/evm_circuit/util.rs [hunk 5]
```diff
@@ -469,6 +469,14 @@ impl<'a> StepRws<'a> {
...
         self.offset += inc
     }
+    /// Decrement the step rw operation offset by `dec`.
+    pub(crate) fn offset_sub(&mut self, dec: usize) {
+        self.offset -= dec
+    }
+    /// Reset the step rw operation offset to 0.
+    pub(crate) fn offset_reset(&mut self) {
+        self.offset = 0
+    }
```
Comment (ChihChengLiang) [inline zkevm-circuits/src/evm_circuit/util.rs:479]: These two APIs are no longer consumed in the codebase.

We shouldn't have API to decrease or rest offset for the reasons discussed in my previous comments.
```suggestion
```

**Review by ChihChengLiang — CHANGES_REQUESTED** [review]
@maschad Great work replacing the manual counting of read-write indices with the automatic one. The assignment looks cleaner now.

I added some comments on fixing the APIs and some style fixes. We can merge the PR after the feedback are addressed.

File: zkevm-circuits/src/evm_circuit/execution/returndatacopy.rs (hunk 4 above)
Comment (maschad) [reply zkevm-circuits/src/evm_circuit/execution/returndatacopy.rs:195]: Thanks! I missed this

**Comment by maschad** [comment]
Thanks for your review @ChihChengLiang I've made the requested adjustments.

**Review by ChihChengLiang — APPROVED** [review]
LGTM. Thanks for the fix

**Pull request merged**